[GenFiller]: Add genFiller module for fiducial measurements

bonanomi opened the pull request:
This PR implements the `genFiller.py` module, inspired from the [`GenFiller.cc`](https://github.com/CJLST/ZZAnalysis/blob/Run2UL_22/AnalysisStep/plugins/GenFiller.cc) plugin used in miniAOD. This module aims at defining the fiducial phase space at gen-level and stores in the final `.root` ntuple the relevant branches for fiducial and differential analyses.

The code at the current stage includes:
* All the selection of objects and definition of the fiducial phase space;
* Store to ntuples the information of the ZZ and Z candidates, together with a `bool` flag for events that pass fiducial selections;

Further developments can target:
* Inclusion of MELA discriminants at gen-level (can be postponed, not relevant for early Run3 analysis nor for any other legacy analysis that is not fiducial xsec measurements);
* Inclusion of gen-level jets (can be postponed, not relevant for early Run3 analysis, especially if we don't have reco-level jets);
* Storage to ntuples of leptons IDs, four-vector etc. This is a trivial development, which does not impact the merging of this PR.

Some plots demonstrating the testing of the code:

![image](https://github.com/CJLST/ZZAnalysis/assets/19261234/2ce9531d-2a4b-4be4-b2f4-324bc2105555)
![image](https://github.com/CJLST/ZZAnalysis/assets/19261234/afc409d4-b1cf-4293-b1c1-8e25f16e2ac0)

bonanomi committed:
[genFiller] Initial commit on genFiller for fiducial analysis

bonanomi committed:
[genfiller] finish gen level selection and preliminary storage to final tree

bonanomi committed:
[genfiller] Filter momId, add comments, save branches for Run3 xsec

namapane commented:
Ciao Matteo,
That's cool! Thanks for taking care of this.
However, some gen info was already collected in module [mcTruthAnalyzer.py](https://github.com/CJLST/ZZAnalysis/blob/Run3/NanoAnalysis/python/mcTruthAnalyzer.py), so first of all I would propose to merge the two modules into a single one.
Also, it would be useful to follow the nanoAOD naming convention for variables, so that variables related the same object start with a common prefix+"`_`". For the gen4l system I choose "`GenZZ_`" as prefix. (I have no strong opinion in changing it to "`Gen4l_`" if you prefer). Following this convention enables to group variables automatically, for example in documentation, and to be able to represent objects in Collections.
I would suggest this naming:

GenDressedLeps_ + pt, RelIso 
GenZZ_ (or Gen4l) + mass, pt, eta, phi, rapidity, Z1mass, Z2mass (*), passedFiducial
 
(*) consistently with the [naming for reco candidates](https://github.com/CJLST/ZZAnalysis/blob/Run3/NanoAnalysis/python/ZZFiller.py#L132)

N.

namapane commented:
A small comment related to the "Storage to ntuples of leptons IDs, four-vector etc": MCTruthAnalyzer already includes storage of indices for the relevant genParts [[here](https://github.com/CJLST/ZZAnalysis/blob/Run3/NanoAnalysis/python/mcTruthAnalyzer.py#L27-L30)]. This is the preferred way to store this information (indices rather than storing a copy of the selected values); helpers can then be provided to manipulate the indicies in a simple way; for example, to retrieve the four leptons of a reco candidate as objects we have [this helper](https://github.com/CJLST/ZZAnalysis/blob/Run3/NanoAnalysis/python/tools.py#L33-L38). A similar one can be implemented to retrieve the gen particle as a list of objects from the indices.

bonanomi commented:
Ciao Nicola,
At first I thought of implementing this module extending the `mcTruthAnalyzer` but then I recalled that also in the miniAOD version of the code we had two separate plugins: `MCTruth` and `GenFiller`. Hence the reason of keeping them separate also here.

For what concerns the name of the observables: `GenDressedLeps_ + pt, RelIso` was more of a test, so these variables can probably be dropped (or renamed in case we figure we need them for the analysis). The `GEN_` variables were named following the same conventions used in the miniAOD code, because changing name of the observables here would imply that we need to change the name of the observables also downstream in the analysis code (even though I fear we will have to change/restyle the analysis code anyway, but I would like to keep the modifications as minimal as possible). However, if there is a strong preference on the names, I can also use the naming you suggest.

For what concerns storing the indices instead of the full object: I agree this is the best practice and in the module I wrote the indices are saved (sorted for the 4l system and additional indices for the extra leptons in the event), so this is an information that can be easily added to the ntuples. The 8 variables I added for the moment (the ones plotted above) are stored directly as floats following the same reasoning of the naming conventions: if we have these variables directly in the ntuples, the downstream code used for the analysis can be made backwards compatible more easily. Instead, if now we store only the indices, then we need to add extra code in the analysis framework to reconstruct Z and ZZ candidates.

My last comment is in the merging of this module with mcTruthAnalyser: unless I am missing something, it seems to me that this `genFiller` includes all the relevant information that is there in `mcTruthAnalyser` (pending the saving to ntuples, but eg the indices are already there and are already sorted, see [here](https://github.com/CJLST/ZZAnalysis/blob/dev_genfiller_nano/NanoAnalysis/python/genFiller.py#L264-273)), with the addition of the fiducial phase space definition.
Maybe it's easier to merge `mcTruthAnalyser` into `genFiller`?

Of course, I am open to discuss any software design and implement the modifications before merging this code into the main branch :)

namapane commented:
> Ciao Nicola, At first I thought of implementing this module extending the `mcTruthAnalyzer` but then I recalled that also in the miniAOD version of the code we had two separate plugins: `MCTruth` and `GenFiller`. Hence the reason of keeping them separate also here.

That is historical, as the two pieces were implemented by different people at different times - it would have made sense to merge them back then, but that did not happen. 
I am really in favor of dropping the burden of historical items in favor of a clean, rational implementation with no duplications. That will pay in simplicity and clarity in the long term.

> For what concerns the name of the observables: `GenDressedLeps_ + pt, RelIso` was more of a test, so these variables can probably be dropped (or renamed in case we figure we need them for the analysis). 

I would keep them after renaming as they are already implemented; we can always remove/comment these variables later if we need to save space. 

> The `GEN_` variables were named following the same conventions used in the miniAOD code, because changing name of the observables here would imply that we need to change the name of the observables also downstream in the analysis code (even though I fear we will have to change/restyle the analysis code anyway, but I would like to keep the modifications as minimal as possible). However, if there is a strong preference on the names, I can also use the naming you suggest.

Yes, I really think that the benefits of following the standard nanoAOD naming convention, and of using consistent naming, outweight other considerations.

> For what concerns storing the indices instead of the full object: I agree this is the best practice and in the module I wrote the indices are saved (sorted for the 4l system and additional indices for the extra leptons in the event), so this is an information that can be easily added to the ntuples. The 8 variables I added for the moment (the ones plotted above) are stored directly as floats following the same reasoning of the naming conventions: if we have these variables directly in the ntuples, the downstream code used for the analysis can be made backwards compatible more easily. Instead, if now we store only the indices, then we need to add extra code in the analysis framework to reconstruct Z and ZZ candidates.

I think there's a misunderstanding here, I did not mean to save the indices only! I was referring to the comment on adding "Storage to ntuples of leptons IDs, four-vector etc." in the PR description. What I meant is that instead of adding the pt, eta, phi, mass, id, ecc of each gen lepton, we should just use the indices.

> My last comment is in the merging of this module with mcTruthAnalyser: unless I am missing something, it seems to me that this `genFiller` includes all the relevant information that is there in `mcTruthAnalyser` (pending the saving to ntuples, but eg the indices are already there and are already sorted, see [here](https://github.com/CJLST/ZZAnalysis/blob/dev_genfiller_nano/NanoAnalysis/python/genFiller.py#L264-273)), with the addition of the fiducial phase space definition. Maybe it's easier to merge `mcTruthAnalyser` into `genFiller`?

I have no problem with the module being called genFiller instead of mcTruthAnalyzer :-) providing that there is no code duplication and that additions don't remvove or change the behaviour for variables already implemented (unless that's for a good reason, of course!).

bonanomi commented:
Hi Nicola,
Thanks for following up so promptly!
I agree with all your points.
I will take care of implementing the changes in the naming conventions and merging the two pieces of code into a single one ( I agree that in the past the splitting was introduced only for historical reasons, but framework-wise it doesn't make much sense to have two independent modules that acts on the same objects - and perform one a subset of operations of the other -): I will work on this in the following days, but at least now we have a baseline skeleton for this part of code that is fundamental for the fiducial analysis.

I will switch the name of this PR to work in progress, so that we can resume the review and/or merge once the development is completed.

bonanomi committed:
[genFiller] Rename variables and save Z cands indices

bonanomi commented:
Hi @namapane ,
Sorry if I get back to this only now.
In my last commit I pushed some modifications aiming at addressing your comments. In particular:
* I have renamed the variables to be consistent with nanoAOD naming conventions;
* I have stored the sorted indices of the leptons composing the Z cands in the `GenZZ_Z*l*Idx` branches;

Note that I have done all the above for events that pass the fiducial cuts of the analysis. Maybe this can be made a bit more general (ie for all the ZZ cands, regardless of whether they fall within the fiducial volume), but that is an info we do not use.

I have also added the `lhe_logger()` method to print the MC history as done in `mcTruthAnalyzer.py`.

I think this code can be used in favor of `mcTruthAnalyzer.py`, as it now saves the variables with a consistent name and can be used for the fiducial analysis, while `mcTruthAnalyzer.py` cannot. 

The Z candidates selection should be equivalent in the two codes, with the addition of `genFiller.py` implementing all the relevant considerations for the fiducial analysis.

The only difference I could spot (and I did not implement because I would like to avoid saving the same variable built in two different ways in the ntuples) is related to the FSR recovery at gen-level. [This](https://github.com/CJLST/ZZAnalysis/blob/dev_genfiller_nano/NanoAnalysis/python/genFiller.py#L75-98) is how it is implemented in the `genFiller`, and it is consistent with what we had in `GenFiller.cc` in the miniAOD analysis (and to what we do in the fiducial analysis). In the `mcTruthAnalyzer.py` this is done in a [slightly different manner](https://github.com/CJLST/ZZAnalysis/blob/dev_genfiller_nano/NanoAnalysis/python/mcTruthAnalyzer.py#L87-96), checking directly the matching with the photons in the `FsrPhoton` collection.
The current implementation in `genFiller` indeed reflects what we used in the past differential analyses: do you think we need the info as stored in the `mcTruthAnalyzer.py` at some point? If so, I could implement these FSR indices in the `genFiller`, otherwise (I don't think this info is used anywhere else but the fiducial analysis) we could probably skip this and merge the code (unless I am missing some obvious addition that can be done before merging).

namapane commented:
Ciao Matteo,
Thanks for following up. Regarding your comments:
The purpose of "FsrPhoton_genFsrIdx" is to get the gen photon matched to each **reconstructed** FRS photon, which is useful for FSR recovery performance studies. It is a different problem of getting dressed gen leptons (where you want to collect gen FSR regardless of whether they are reconstructed or not). So I think each implementation should they should stay as is.   

I think we should keep gen indices etc. also for events that do not pass fiducial cuts, because we may want to use the same code for many other studies besisdes the fiducial analysis (imagine one wants to make gen-level plots or debugging of some event for whatever reason, (s)he would have to figure out why the information is missing) As I would like to merge the two modules into one (and import GenZZ_FinalState and FsrPhoton_genFsrIdx in genFiller.py) I can also take care of this.

bonanomi committed:
[genFiller] Fix Z cands indices

bonanomi committed:
[genFiller] Rename branches and move to tools.py util functions common to mcTruthAnalyzer.py
